x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/parser/ExpressionBuilder.java

```diff
@@ -73,8 +74,8 @@ public Expression visitSingleExpression(EqlBaseParser.SingleExpressionContext ct
     }
 
     @Override
-    public List<Expression> visitJoinKeys(JoinKeysContext ctx) {
-        return ctx != null ? expressions(ctx.expression()) : emptyList();
+    public List<Attribute> visitJoinKeys(JoinKeysContext ctx) {
```

Comment:
Since this doesn't precisely mirror the grammar, what errors will it cause if you do this?
```sql
by substring(field)
```

I completely agree with this restriction for Elasticsearch, even though Endpoint isn't bound by it. Will this raise a good error message or does it get buried in an uncaught exception?

Reply:
Right now it will throw an exception since it expects a materialized expression which will fail for a function.
This needs improving but I'd rather handle that in a follow-up PR.